File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/operation/meta/schema/write/AttSchemaGqlWriter.java
package ru.citeck.ecos.records3.record.operation.meta.schema.write;

import ru.citeck.ecos.commons.utils.NameUtils;
import ru.citeck.ecos.records3.RecordConstants;
import ru.citeck.ecos.records3.record.operation.meta.schema.SchemaAtt;

import java.util.List;

/**
 * Writer to convert schema to legacy GraphQL format.
 */
public class AttSchemaGqlWriter implements AttSchemaWriter {

    @Override
    public void write(SchemaAtt attribute, StringBuilder sb) {
        sb.append(".");
        writeInner(attribute, sb);
    }

    private void writeInner(SchemaAtt attribute, StringBuilder sb) {

        String alias = attribute.getAlias();
        String name = attribute.getName();
        List<SchemaAtt> inner = attribute.getInner();

        if (sb.length() > 1 && !alias.isEmpty()) {
            sb.append(NameUtils.escape(alias)).append(":");
        }

        if (attribute.isScalar()) {
            sb.append(name, 1, name.length());
            return;
        }

        if (name.charAt(0) == '_' && inner.size() == 1) {
            if (name.equals(RecordConstants.ATT_EDGE)) {
                writeEdge(attribute, sb);
                return;
            }
            if (name.equals(RecordConstants.ATT_AS)) {
                writeAs(attribute, sb);
                return;
            }
            if (name.equals(RecordConstants.ATT_HAS)) {
                writeHas(attribute, sb);
                return;
            }
        }

        sb.append("att");
        if (attribute.isMultiple()) {
            sb.append("s");
        }
        sb.append("(n:\"")
            .append(name.replace("\"", "\\\""))
            .append("\"){");

        for (int i = 0; i < inner.size(); i++) {
            writeInner(inner.get(i), sb);
            if (i < inner.size() - 1) {
                sb.append(",");
            }
        }
        sb.append("}");
    }

    private void writeEdge(SchemaAtt attribute, StringBuilder sb) {

        SchemaAtt innerAtt = attribute.getInner().get(0);

        sb.append("edge(n:\"")
            .append(innerAtt.getName().replace("\"", "\\\""))
            .append("\"){");

        for (SchemaAtt att : innerAtt.getInner()) {
            writeEdgeInner(att, sb);
            sb.append(",");
        }
        if (innerAtt.getInner().size() > 0) {
            sb.setLength(sb.length() - 1);
        }

        sb.append("}");
    }

    private void writeEdgeInner(SchemaAtt attribute, StringBuilder sb) {

        String alias = attribute.getAlias();
        if (!alias.isEmpty()) {
            sb.append(NameUtils.unescape(alias)).append(":");
        }
        sb.append(attribute.getName());

        switch (attribute.getName()) {
            case "val":
            case "vals":
            case "options":
            case "distinct":
            case "createVariants":
                sb.append("{");
                for (SchemaAtt att : attribute.getInner()) {
                    writeInner(att, sb);
                    sb.append(",");
                }
                if (!attribute.getInner().isEmpty()) {
                    sb.setLength(sb.length() - 1);
                }
                sb.append("}");
        }
    }

    private void writeAs(SchemaAtt attribute, StringBuilder sb) {

        SchemaAtt innerAtt = attribute.getInner().get(0);

        sb.append("as(n:\"")
            .append(innerAtt.getName().replace("\"", "\\\""))
            .append("\"){");

        for (SchemaAtt att : innerAtt.getInner()) {
            writeInner(att, sb);
        }

        sb.append("}");
    }

    private void writeHas(SchemaAtt attribute, StringBuilder sb) {
        sb.append("has(n:\"")
            .append(attribute.getInner().get(0).getName().replace("\"", "\\\""))
            .append("\")");;
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/schema/write/AttSchemaGqlWriter.kt
package ru.citeck.ecos.records3.record.op.atts.service.schema.write

import ru.citeck.ecos.commons.utils.NameUtils
import ru.citeck.ecos.records2.RecordConstants
import ru.citeck.ecos.records3.record.op.atts.service.schema.SchemaAtt

/**
 * Writer to convert schema to legacy GraphQL format.
 */
class AttSchemaGqlWriter : AttSchemaWriter {

    override fun write(attribute: SchemaAtt, out: StringBuilder) {
        out.append(".")
        writeInner(attribute, out)
    }

    private fun writeInner(attribute: SchemaAtt, sb: StringBuilder) {
        writeInnerAtt(attribute, sb)
        for (processor in attribute.processors) {
            sb.append("|").append(processor.type).append("(")
            val argsSize = processor.arguments.size
            for (i in 0 until argsSize) {
                sb.append(processor.arguments[i].toString())
                if (i < argsSize - 1) {
                    sb.append(',')
                }
            }
            sb.append(")")
        }
    }

    private fun writeInnerAtt(attribute: SchemaAtt, sb: StringBuilder) {

        val alias: String = attribute.alias
        val name: String = attribute.name

        if (sb.length > 1 && alias.isNotEmpty()) {
            sb.append(NameUtils.escape(alias)).append(":")
        }
        if (attribute.isScalar()) {
            sb.append(name, 1, name.length)
            return
        }
        if (name[0] == '_' && attribute.inner.size == 1) {
            if (name == RecordConstants.ATT_EDGE) {
                writeEdge(attribute, sb)
                return
            }
            if (name == RecordConstants.ATT_AS) {
                writeAs(attribute, sb)
                return
            }
            if (name == RecordConstants.ATT_HAS) {
                writeHas(attribute, sb)
                return
            }
        }
        sb.append("att")
        if (attribute.multiple) {
            sb.append("s")
        }
        sb.append("(n:\"")
            .append(name.replace("\"", "\\\""))
            .append("\"){")

        val inner = attribute.inner

        for (i in inner.indices) {
            writeInner(inner[i], sb)
            if (i < inner.size - 1) {
                sb.append(",")
            }
        }
        sb.append("}")
    }

    private fun writeEdge(attribute: SchemaAtt, sb: StringBuilder) {

        val innerAtt: SchemaAtt = attribute.inner[0]

        sb.append("edge(n:\"")
            .append(innerAtt.name.replace("\"", "\\\""))
            .append("\"){")

        for (att in innerAtt.inner) {
            writeEdgeInner(att, sb)
            sb.append(",")
        }

        if (innerAtt.inner.isNotEmpty()) {
            sb.setLength(sb.length - 1)
        }

        sb.append("}")
    }

    private fun writeEdgeInner(attribute: SchemaAtt, sb: StringBuilder) {

        val alias: String = attribute.alias

        if (alias.isNotEmpty()) {
            sb.append(NameUtils.escape(alias)).append(":")
        }

        sb.append(attribute.name)

        when (attribute.name) {
            "val",
            "vals",
            "options",
            "distinct",
            "createVariants" -> {
                sb.append("{")
                for (att in attribute.inner) {
                    writeInner(att, sb)
                    sb.append(",")
                }
                if (attribute.inner.isNotEmpty()) {
                    sb.setLength(sb.length - 1)
                }
                sb.append("}")
            }
        }
    }

    private fun writeAs(attribute: SchemaAtt, sb: StringBuilder) {
        val innerAtt: SchemaAtt = attribute.inner[0]
        sb.append("as(n:\"")
            .append(innerAtt.name.replace("\"", "\\\""))
            .append("\"){")
        for (att in innerAtt.inner) {
            writeInner(att, sb)
        }
        sb.append("}")
    }

    private fun writeHas(attribute: SchemaAtt, sb: StringBuilder) {
        sb.append("has(n:\"")
            .append(attribute.inner[0].name.replace("\"", "\\\""))
            .append("\")")
    }
}
